### PR TITLE
ARROW-491: [Format / C++] Add FixedWidthBinary type to format, C++ implementation

### PR DESCRIPTION
**Pull request by wesm**
I have a bunch more work to do on the C++ implementation:

- [x] Test builder class
- [x] Test array API (slice, etc.)
- [x] Implement/test ArrayEquals/ArrayRangeEquals
- [x] Implement `PrettyPrint` (may want to encode to hexadecimal, I don't think that BinaryArray prints properly right now for non-ASCII/UTF8 data)
- [x] Add IPC roundtrip tests

In the meantime, @julienledem @nongli or others could you look at the changes to the format Flatbuffers and let me know if that looks right to you? Thanks

### CODE REVIEW

**Comment by nongli** [comment]
flatbuf changes look simple enough and good with me.

**Comment by wesm** [comment]
OK, this is ready for review

File: cpp/src/arrow/builder.cc [hunk 1]
```diff
@@ -121,6 +121,14 @@ void ArrayBuilder::UnsafeAppendToBitmap(const uint8_t* valid_bytes, int64_t leng
   uint8_t bitset = null_bitmap_data_[byte_offset];
 
   for (int64_t i = 0; i < length; ++i) {
+    if (bit_offset == 8) {
```
Comment (xhochy) [inline cpp/src/arrow/builder.cc]: Can you give me a hint what the difference in behaviour is?

**Review by xhochy — APPROVED** [review]
LGTM except that I don't understand how the bitsetting changed. I fail to see a different bahaviour.

File: cpp/src/arrow/builder.cc (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/builder.cc]: When the length is a power of 2, on the last iteration of the loop, a value past the end of the null bitmap is accessed, causing valgrind to complain. I only stumbled on this by accident by using 4096 as the length in my test case

Comment (wesm) [inline cpp/src/arrow/builder.cc]: This moves the `bit_offset++` after this if block

Comment (xhochy) [inline cpp/src/arrow/builder.cc]: Thanks, now I get it :)

**Comment by wesm** [comment]
green build: https://travis-ci.org/wesm/arrow, except for the stalled OS X build because this isn't rebased on the nomkl patch

File: cpp/src/arrow/builder.cc (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/builder.cc]: Might want to look at other places where this fast-bitsetting logic exists to see if there might be an inappropriate memory access

File: format/Message.fbs [hunk 2]
```diff
@@ -68,6 +68,11 @@ table Utf8 {
 table Binary {
 }
 
+table FixedWidthBinary {
+  /// Number of bytes per value
+  byteWidth: int;
+}
```
Comment (julienledem) [inline format/Message.fbs]: do we want an equivalent FixedWidthUtf8?

**Review by julienledem — COMMENTED** [review]
from the metadata perspective, this looks good to me.

File: format/Message.fbs (hunk 2 above)
Comment (wesm) [inline format/Message.fbs]: Utf8 code points use a variable number of bytes (1 to 4), so to fully account for fixed width unicode, you would have to store (N code points) * 4 bytes. This is likely to be wasteful compared with the current list-like / variable-length UTF8 representation

**Pull request closed without merging**